### ISSUE TITLE
Bug: not 128bit aligned with `repr-c` feature

### ISSUE THREAD

**Issue by Erigara** [paupino/rust-decimal#649]
I've noticed that while [docs](https://docs.rs/rust_decimal/1.31.0/rust_decimal/index.html#c-repr) claim that `Decimal` is 128bit aligned it's not the case which could lead to UB ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e32abe7612da2f558fed71ab7cc19288) run with miri).

Either docs should be edited of fix implemented. 

I can work on the fix by introducing new ZST field `_align: [u128; 0]` into `Decimal` which would force proper alignment of the type to be equal to `u128`.

```rust
#[cfg_attr(feature = "c-repr", repr(C))]
pub struct Decimal {
    // Bits 0-15: unused
    // Bits 16-23: Contains "e", a value between 0-28 that indicates the scale
    // Bits 24-30: unused
    // Bit 31: the sign of the Decimal value, 0 meaning positive and 1 meaning negative.
    flags: u32,
    // The lo, mid, hi, and flags fields contain the representation of the
    // Decimal value as a 96-bit integer.
    hi: u32,
    lo: u32,
    mid: u32,
    // Enforce proper alignment
    #[cfg(feature = "c-repr")]
    _align: [u128; 0],
}
```

**Comment by Tony-Samuels** [comment]
I would *suggest* the approach should be to update the docs to remove this guarantee. The main problem is that whilst you're correct that the current implementation is not aligned to 128 bits, your proposed implementation won't be either!

You can see this in your playground link where the error if the alignments don't match displays that `u128` is 8 byte aligned (i.e. only 64 bits). 

So we would have two options for fixing the alignment, but both break the principle of least surprise, imo:
1. `repr(c, align(16))` 
    * `+` Aligns to 128 bits
    * `-` The alignment on some platforms will be greater than than of `u128`
2. `_align: [u128; 0]`
    * `+` The alignment across all platforms will match `u128`'s
    * `-` The alignment is not guaranteed to be 128 bits

**Comment by Tony-Samuels** [comment]
A separate issue to removing the docs is still whether the maintainer wishes to change the alignment to match that of `u128`, which wouldn't be unreasonable. However I think the doc need to be updated to fix this issue regardless.

**Comment by Erigara** [comment]
> You can see this in your playground link if you run it normally (i.e. not with miri) where the error if the alignments don't match displays that u128 is 8 byte aligned (i.e. only 64 bits).

Yeah, i've tried to use `repr(c, align(16))` initially to align to `128bits`, but i forgot that `u128` alignment is tricky.

**Comment by SmnTin** [comment]
@Tony-Samuels  
>     1. `repr(c, align(16))`
>        
>        * `+` Aligns to 128 bits
>        * `-` The alignment on some platforms will be greater than than of `u128`

Yes, Rust only guarantees that the alignment is at least 16 bytes. But could you please provide an example of a platform where the alignment will be greater? As far as I know, on all supported platforms, the alignment will be exactly 16 bytes.

I think the only actual concern is that, because the alignment hasn't been enforced for some time, fixing it now might be a breaking change for some users.

**Comment by Tony-Samuels** [comment]
https://releases.rs/docs/1.77.0/

Looks like this was actually fixed just a few weeks later. We should be able to add proper alignment now.

**Comment by SmnTin** [comment]
I agree that the 3rd option sounds most reasonable. It makes it a backwards-compatible change.

The main application is to be able to `transmute` a decimal to and back from `u128`. This might be useful to pass it in the FFI as `u128` without exposing the type definition. I don't have any more concrete and realistic examples yet, however 😅
